### PR TITLE
[cymulate-205] Cymulate Immediate Threats Integration

### PR DESCRIPTION
**Pull request by maor-b**
Initial integration with Cymulate. Supports Immediate Threats.

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Initial integration with Cymulate. Supports Immediate Threats.

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [x] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by maor-b** [comment]
Still work in progress, missing:
- API limit for number of results
- Added CymulateImmediateThreatStatus field and layout
- Smaller playbook changes (Conditions based on Incident.CustomFields.ImmediateThreatStatus etc vs labels)

**Comment by maor-b** [comment]
Hi all we are ready with the requirement
Thanks

**Comment by maor-b** [comment]
Hi @yuvalbenshalom @fvigo , we still waiting for update
there is any updates?

Thanks,
Maor

**Comment by maor-b** [comment]
> Hi @yuvalbenshalom @fvigo , we still waiting for update
> there is any updates?
> 
> Thanks,
> Maor

hi anyone? @fvigo @yuvalbenshalom

**Comment by anara123** [comment]
@maor-b Hello. Thank you for the contribution.
I will review your PR soon.

**Comment by maor-b** [comment]
> @maor-b Hello. Thank you for the contribution.
> I will review your PR soon.

Hi Anar thanks for you replay
can you update please?
Thanks

File: IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Status.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{
+	"associatedToAll": false,
+	"associatedTypes": [
```
Comment (anara123) [inline IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Status.json]: these fields are incident fields, move them under `IncidentFields` folder
Same for all the other fields

File: IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Vector.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"group": 0,
+	"hidden": false,
+	"id": "incident_cymulateimmediatethreatsvector",
+	"isReadOnly": false,
+	"locked": false,
+	"name": "Cymulate Immediate Threats Vector",
+	"neverSetAsRequired": false,
+	"ownerOnly": false,
+	"placeholder": "",
+	"required": false,
```
Comment (anara123) [inline IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Vector.json]: for all the json files add `"fromVersion": "5.0.0"` field

Comment (anara123) [reply IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Vector.json]: all the incident types and incident fields,

File: Integrations/Cymulate/Cymulate.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    - default: false
+      description: The Cymulate module_type
+      isArray: false
+      name: module_type
+      required: true
+      secret: false
+    deprecated: false
+    description: This commands return full cymulate's incidents information
+    execution: false
+    name: cymulate_get_incident_info
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.yml]: change the name of the command to 
`cymulate-get-incident-info` update in the py code and README too

File: Integrations/Cymulate/Cymulate.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    outputs:
+    - contextPath: Cymulate.Incident.Payload
+      description: The Cymulatepayload that generated this incident
+      type: String
+    - contextPath: Cymulate.Incident.Name
+      description: The name of the incident
+      type: String
+    - contextPath: Cymulate.Incident.Status
+      description: The attack status
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.yml]: Status is an Enum. Add description to all the enum, which options exists for status.

File: Integrations/Cymulate/Cymulate.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  commands:
+  - arguments:
+    - default: false
+      description: The incident unique identifier
+      isArray: false
+      name: incident_id
+      required: true
+      secret: false
+    - default: false
+      description: The Cymulate module_type
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.yml]: are there predefined types? if there are then we have single select input

File: Integrations/Cymulate/Cymulate.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  type: 8
+- display: Incident type
+  name: incidentType
+  required: false
+  type: 13
+- display: Show only 'penatrated' incidents
+  hidden: false
+  name: onlyPenatrated
+  required: false
+  type: 8
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.yml]: in the code you use param `module_type` but you haven't defined it in the yml. please define as a single select and define all the available options for module_type

File: Integrations/Cymulate/Cymulate.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    LOG(f'Command being called is: {demisto.command()}')
+    try:
+        client = Client(
+            base_url=BASE_URL,
+            headers=header,
+            verify=False)
+
+        if demisto.command() == 'test-module':
+            # This is the call made when pressing the integration Test button.
+            cymulate_test(client=client)
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: if `isFetch` is enabled then the test should verify that fetch is working. It means you should call the same function that you are calling in the fetch_incidents

File: Integrations/Cymulate/Cymulate.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+                        'ID': incident['ID'],
+                        'Name': incident['Name'],
+                        'Status': incident['Status'],
+                        'Attack Type': incident['Attack_Type'],
+                        'Attack Vector': incident['Attack_Vector'],
+                        'Timestamp': incident['Timestamp']
+                    })
+
+                # Convert to human reading format
+                cymulate_output = tableToMarkdown('Cymulate Resutls', parsed_technical_data, ['ID', 'Name',
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]:
```suggestion
                cymulate_output = tableToMarkdown('Cymulate Results', parsed_technical_data, ['ID', 'Name',
```

File: Integrations/Cymulate/Cymulate.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+                                                                                              'Status', 'Attack Type',
+                                                                                              'Attack Vector',
+                                                                                              'Timestamp'])
+                outputs = {
+                    'data': technical_data
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: I think `data` should be replaced with `Cymulate.Incident(val.ID == obj.ID)`

File: Integrations/Cymulate/Cymulate.yml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  required: true
+  type: 0
+- display: Fetch incidents
+  name: isFetch
+  required: false
+  type: 8
+- display: Incident type
+  name: incidentType
+  required: false
+  type: 13
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.yml]: add 2 parameters that we add to all the integrations.
```
- display: Trust any certificate (not secure)
  name: insecure
  required: false
  type: 8
- display: Use system proxy settings
  name: proxy
  required: false
  type: 8
```

Comment (anara123) [reply Integrations/Cymulate/Cymulate.yml]: use both of them in the code.
insecure should be be passed to `_http_request` to `verify` argument. It verifies the server certificate (allowing the user to choose the secure version)

File: Integrations/Cymulate/Cymulate.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+
+    # Amount limit of returned incidents
+    fetch_limit = 25
+
+    LOG(f'Command being called is: {demisto.command()}')
+    try:
+        client = Client(
+            base_url=BASE_URL,
+            headers=header,
+            verify=False)
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: use the `insecure` param to allow the user to choose.
`verify=False` is insecure option

File: Integrations/Cymulate/Cymulate.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+    if not start_query_time:
+        start_query_time, _ = parse_date_range(first_fetch_time, date_format=DATE_FORMAT, utc=True)
+        start_query_time = ts_add_minutes(start_query_time, 2)
+
+    incidents = []
+
+    # Get the incident from server
+    items = client.get_attack_info(module_type, start_query_time)
+
+    demisto.info("******CYMULATE-LOGS*******")
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: change tthe logs to be `demisto.debug` instead of info

File: Integrations/Cymulate/Cymulate.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+        elif module_type == CymulateModuleTypeEnum.DLP:
+            technical_route = 'dlp/attack/technical'
+        elif module_type == CymulateModuleTypeEnum.BROWSING:
+            technical_route = 'browsing/attack/technical'
+        elif module_type == CymulateModuleTypeEnum.EDR:
+            technical_route = 'edr/attack/technical'
+        elif module_type == CymulateModuleTypeEnum.MITRE:
+            technical_route = 'mitre/attack/technical'
+        elif module_type == CymulateModuleTypeEnum.MAIL:
+            technical_route = 'mail/attack/technical'
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: all these seems to be different routes, are they return different response structure?

File: Integrations/Cymulate/Cymulate.py [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+        datetime: time right now
+    """
+    return datetime.now()
+
+
+def ts_add_minutes(ts, minutes):
+    return (datetime.strptime(ts, DATE_FORMAT) + timedelta(minutes=minutes)).strftime(DATE_FORMAT)
+
+
+def get_fetch_times(last_fetch):
```
Comment (anara123) [inline Integrations/Cymulate/Cymulate.py]: this function not used

File: Layouts/layout-details-Cymulate_Immediate_Threats-V2.json [hunk 15]
```diff
@@ -0,0 +1,341 @@
+{
```
Comment (anara123) [inline Layouts/layout-details-Cymulate_Immediate_Threats-V2.json]:
```json
{
	"TypeName": "Cymulate Immediate Threats",
	"id": "Cymulate Immediate Threats",
	"kind": "details",
	"fromVersion": "5.0.0",
	"layout": {MOVE YOUR JSON HERE},
	"typeId": "Cymulate Immediate Threats",
	"version": -1
}
```
need to change the structure here.

File: IncidentTypes/incidenttype-Cymulate_Immediate_Threats.json [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+	"days": 0,
+	"daysR": 0,
+	"default": false,
+	"disabled": false,
+	"hours": 0,
+	"hoursR": 0,
+	"id": "Cymulate Immediate Threats",
+	"locked": false,
+	"name": "Cymulate Immediate Threats",
+	"playbookId": "75f5e4a9-49ea-4cd1-855b-8f8ba0c46e37",
```
Comment (anara123) [inline IncidentTypes/incidenttype-Cymulate_Immediate_Threats.json]:
```suggestion
	"playbookId": "Cymulate Immediate Threats",
```

File: IncidentTypes/incidenttype-Immediate_Threat.json [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+	"days": 0,
+	"daysR": 0,
+	"default": false,
+	"disabled": false,
+	"hours": 0,
+	"hoursR": 0,
+	"id": "Immediate Threat",
+	"locked": false,
+	"name": "Immediate Threat",
+	"playbookId": "75f5e4a9-49ea-4cd1-855b-8f8ba0c46e37",
```
Comment (anara123) [inline IncidentTypes/incidenttype-Immediate_Threat.json]:
```suggestion
	"playbookId": "Cymulate Immediate Threats",
```

File: TestPlaybooks/playbook-playbook-CymulateTest.yml [hunk 18]
```diff
@@ -0,0 +1,380 @@
+id: playbook-CymulateTest
```
Comment (anara123) [inline TestPlaybooks/playbook-playbook-CymulateTest.yml]: change the name to be just `playbook-CymulateTest.yml`
The name of the test to be `CymulateTest`

File: TestPlaybooks/playbook-playbook-CymulateTest.yml [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+      '#none#':
+      - "9"
+    note: false
+    separatecontext: false
+    task:
+      brand: ""
+      id: 3134025d-89bc-4841-8669-8ee43bf155bf
+      iscommand: true
+      name: fetch_dlp_incidents
+      script: '|||fetch_dlp_incidents'
```
Comment (anara123) [inline TestPlaybooks/playbook-playbook-CymulateTest.yml]: there are no such commands in your integration
```
fetch_dlp_incidents
fetch_hopper_incidents
fetch_waf_incidents
```

**Comment by anara123** [comment]
@maor-b Hey. After the PR is ready, let me know and we will schedule a session for demoing the integration

**Comment by maor-b** [comment]
> @maor-b Hey. After the PR is ready, let me know and we will schedule a session for demoing the integration

Hi Anar , can we talk on slack?

**Comment by maor-b** [comment]
> Missing README.md

README.md exists in cymulate folder

File: IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Vector.json (hunk 2 above)
Comment (maor-b) [reply IndicatorFields/incidentfield-Cymulate_Immediate_Threats_Vector.json]: added

**Comment by anara123** [comment]
@maor-b sure, you can contact me on slack. I am in DFIR channel

File: Integrations/Cymulate/Cymulate.yml (hunk 5 above)
Comment (maor-b) [reply Integrations/Cymulate/Cymulate.yml]: no, currently only static predefine type

File: Layouts/layout-details-Cymulate_Immediate_Threats-V2.json (hunk 15 above)
Comment (maor-b) [reply Layouts/layout-details-Cymulate_Immediate_Threats-V2.json]: paste the json to the "layout" prop

File: TestPlaybooks/playbook-playbook-CymulateTest.yml (hunk 19 above)
Comment (maor-b) [reply TestPlaybooks/playbook-playbook-CymulateTest.yml]: update to the right yaml

**Comment by maor-b** [comment]
Hi Anar, thanks for your replay and review

I fixed all your comment

please update when we can schedule a demo

thanks

File: IncidentTypes/incidenttype-Cymulate_Immediate_Threats.json [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+	"days": 0,
+	"daysR": 0,
+	"default": false,
+	"disabled": false,
+	"hours": 0,
+	"hoursR": 0,
+	"id": "Cymulate Immediate Threats",
+	"locked": false,
+	"name": "Cymulate Immediate Threats",
+	"playbookId": "Cymulate Immediate Threats",
```
Comment (anara123) [inline IncidentTypes/incidenttype-Cymulate_Immediate_Threats.json]: missing fromVersion

File: IncidentTypes/incidenttype-Immediate_Threat.json [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+	"preProcessingScript": "",
+	"readonly": false,
+	"reputationCalc": 0,
+	"sla": 168,
```
Comment (anara123) [inline IncidentTypes/incidenttype-Immediate_Threat.json]: missing fromVersion

File: Integrations/Cymulate/Cymulate.yml (hunk 5 above)
Comment (anara123) [reply Integrations/Cymulate/Cymulate.yml]: what do you mean by static predefine type?
I mean it seems that there are predefine list of possible values that module_type can be:
```
	class CymulateModuleTypeEnum(enum.Enum):
    """
       Enum class of module types of cymulate attacks
    """

    APT = '1',
    IMMEDIATE_THREATS = '2',
    HOPPER = '3',
    PHISHING = '4',
    WAF = '5',
    DLP = '6',
    BROWSING = '7',
    EDR = '8',
    MITRE = '9',
    MAIL = '10'
```

File: Integrations/Cymulate/Cymulate.yml [hunk 22]
```diff
@@ -0,0 +1,67 @@
+category: Vulnerability Management
+commonfields:
+  id: Cymulate
+  version: -1
+configuration:
+- defaultvalue: 179743a665726a99fdee3bb0d10ba3fe
+  display: x-token
```
[the remark this answers is not in the document]
Comment (maor-b) [reply Integrations/Cymulate/Cymulate.yml]: sent in email

File: Integrations/Cymulate/Cymulate.yml (hunk 6 above)
Comment (maor-b) [reply Integrations/Cymulate/Cymulate.yml]: as we agreed with francesco - we now support only module_type="immediate_threats"

File: Integrations/Cymulate/Cymulate.py (hunk 13 above)
Comment (maor-b) [reply Integrations/Cymulate/Cymulate.py]: now we support only 'immediate-threats'

**Pull request merged**